File: VB/CustomTokenEdit/RepositoryItemCustomTokenEdit.vb
Imports DevExpress.Skins
Imports DevExpress.Utils
Imports DevExpress.XtraEditors
Imports DevExpress.XtraEditors.Registrator
Imports DevExpress.XtraEditors.Repository
Imports System.Reflection

Namespace Example.CustomTokenEdit
    <UserRepositoryItem("Register")>
    Public Class RepositoryItemCustomTokenEdit
        Inherits RepositoryItemTokenEdit

        Shared Sub New()
            Register()
        End Sub
        Public Sub New()
            checkBoxCollection_Renamed = GetCheckBoxImages()
        End Sub

        Friend Const EditorName As String = "CustomTokenEdit"
        Public Shared Sub Register()
            Dim img As Image = Nothing
            Try
                img = Bitmap.FromStream(Assembly.GetExecutingAssembly().GetManifestResourceStream("DevExpress.CustomEditors.CustomEdit.bmp"))
            Catch
            End Try
            EditorRegistrationInfo.Default.Editors.Add(New EditorClassInfo(EditorName, GetType(CustomTokenEdit), GetType(RepositoryItemCustomTokenEdit), GetType(CustomTokenEditViewInfo), New CustomTokenEditPainter(), True, img))
        End Sub
        Public Overrides ReadOnly Property EditorTypeName() As String
            Get
                Return EditorName
            End Get
        End Property

        'INSTANT VB NOTE: The variable checkBoxCollection was renamed since Visual Basic does not allow variables and other class members to have the same name:
        Private ReadOnly checkBoxCollection_Renamed As ImageCollection
        Public ReadOnly Property CheckBoxCollection() As ImageCollection
            Get
                Return checkBoxCollection_Renamed
            End Get
        End Property

        Protected Overridable Function GetCheckBoxImages() As ImageCollection
            Dim skin As Skin = EditorsSkins.GetSkin(DevExpress.LookAndFeel.UserLookAndFeel.Default)
            Dim skinElement As SkinElement = skin("CheckBox")
            If skinElement Is Nothing Then
                Return Nothing
            End If
            Return skinElement.Image.GetImages()
        End Function

        Protected Overrides Function CreateValidationController() As TokenEditValidationController
            Return New CustomTokenEditValidationController(Me)
        End Function

        Protected Overrides Function CreateCheckedItemCollection() As TokenEditCheckedItemCollection
            Return New CustomTokenEditCheckedItemCollection(Me)
        End Function

        Protected Overrides Function CreateSelectedItemCollection() As TokenEditSelectedItemCollection
            Return New CustomTokenEditSelectedItemCollection(Me)
        End Function

        Public Event TokenCheckedChanged As EventHandler(Of TokenCheckedChangedEventArgs)
        Protected Friend Sub OnTokenCheckedChanged(ByVal token As CustomTokenEditToken, ByVal repaint As Boolean)
            RepaintEditIfRequired(repaint)
            RaiseTokenCheckedChanged(token)
        End Sub

        Private Sub RaiseTokenCheckedChanged(ByVal token As CustomTokenEditToken)
            Dim handler As EventHandler(Of TokenCheckedChangedEventArgs) = TokenCheckedChangedEvent
            If handler IsNot Nothing Then
                TokenCheckedChangedEvent(Me, New TokenCheckedChangedEventArgs(token))
            End If
        End Sub

        Private Sub RepaintEditIfRequired(ByVal repaint As Boolean)
            Dim edit As CustomTokenEdit = TryCast(OwnerEdit, CustomTokenEdit)
            If repaint AndAlso edit IsNot Nothing Then
                edit.Repaint()
            End If
        End Sub
    End Class
End Namespace

File: CS/TokenEdit_Glyph_InEndOfToken/CustomTokenEdit/RepositoryItemCustomTokenEdit.cs
using DevExpress.Skins;
using DevExpress.Utils;
using DevExpress.XtraEditors;
using DevExpress.XtraEditors.Registrator;
using DevExpress.XtraEditors.Repository;
using System;
using System.Collections.Generic;
using System.Linq;

namespace TokenEdit_Glyph_InEndOfToken.CustomTokenEdit {
    [UserRepositoryItem("Register")]
    public class RepositoryItemCustomTokenEdit : RepositoryItemTokenEdit {
        static RepositoryItemCustomTokenEdit() { Register(); }
        public RepositoryItemCustomTokenEdit() { checkBoxCollection = GetCheckBoxImages(); }

        internal const string EditorName = "CustomTokenEdit";
        public static void Register() {
            EditorRegistrationInfo.Default.Editors.Add(new EditorClassInfo(EditorName, typeof(CustomTokenEdit),
                typeof(RepositoryItemCustomTokenEdit), typeof(CustomTokenEditViewInfo),
                new CustomTokenEditPainter(), true, null));
        }
        public override string EditorTypeName { get { return EditorName; } }

        readonly SkinElement checkBoxCollection;
        public SkinElement CheckBoxSkinElement { get { return checkBoxCollection; } }

        protected virtual SkinElement GetCheckBoxImages() {
            Skin skin = EditorsSkins.GetSkin(DevExpress.LookAndFeel.UserLookAndFeel.Default);
            SkinElement skinElement = skin["CheckBox"];
            if (skinElement == null)
                return null;
            return skinElement;
        }

        protected override TokenEditCheckedItemCollection CreateCheckedItemCollection() {
            return new CustomTokenEditCheckedItemCollection(this);
        }

        protected override TokenEditSelectedItemCollection CreateSelectedItemCollection() {
            return new CustomTokenEditSelectedItemCollection(this);
        }

        protected override TokenEditToken CreateTokenInstance(string description, object value, bool autoPopulated) {
            CustomTokenEditToken token = new CustomTokenEditToken(description, value, this as RepositoryItemCustomTokenEdit);
            return token;
        }

        public event EventHandler<TokenCheckedChangedEventArgs> TokenCheckedChanged;
        protected internal void OnTokenCheckedChanged(CustomTokenEditToken token, bool repaint) {
            RepaintEditIfRequired(repaint);
            RaiseTokenCheckedChanged(token);
        }

        private void RaiseTokenCheckedChanged(CustomTokenEditToken token) {
            EventHandler<TokenCheckedChangedEventArgs> handler = TokenCheckedChanged;
            if (handler != null)
                handler(this, new TokenCheckedChangedEventArgs(token));
        }

        private void RepaintEditIfRequired(bool repaint) {
            CustomTokenEdit edit = OwnerEdit as CustomTokenEdit;
            if (repaint && edit != null)
                edit.Repaint();
        }
    }
}
